USB Host: integrate class API [3: UVC]

josuah opened the pull request:
Dependency:

- https://github.com/zephyrproject-rtos/zephyr/pull/94590
- https://github.com/zephyrproject-rtos/zephyr/pull/103369

Downstream:

- https://github.com/zephyrproject-rtos/zephyr/pull/94085
- https://github.com/zephyrproject-rtos/zephyr/pull/99775

**NOTE: this PR also contains the commit of #94504 and #94590**

On top of the commits from the previous PRs:

- split UVC helpers out of the device stack so that they can be used with the host stack.

UVC Device still works:

```
west build --build-dir build-uvc -b nrf52840dk/nrf52840 -S video-sw-generator samples/subsys/usb/uvc/
```

![mpv-shot0001](https://github.com/user-attachments/assets/09162a44-6acf-4568-b1c7-e3fe0f5946cb)

All tests are done on top of https://github.com/zephyrproject-rtos/zephyr/compare/main...josuah:zephyr:pr_usb_host_class_api3

Commit:
tests: usb: uvc: connect USB device and host via UVB

Add a test to run the USB Video Class host support by using
the existing Zephyr USB Video Class device support.
This allows running implementing the host side from the device side.
A draft implementation of UVC is added leveraging this test.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: uvc: placeholder implementation

Loop through each of the VideoStreaming and VideoControl descriptor
to parse them. This is meant as a stub for the purpose of testing the
class API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

AidenHu commented:
> Dependency:
> 
>     * [USB Host: integrate class API [2: helpers] #94590](https://github.com/zephyrproject-rtos/zephyr/pull/94590)
> 
> 
> Downstream:
> 
>     * [usb: host: class: support for usb host video class #94085](https://github.com/zephyrproject-rtos/zephyr/pull/94085)
> 
> 
> **NOTE: this PR also contains the commit of #94504 and #94590**
> 
> On top of the commits from the previous PRs:
> 
>     * split UVC helpers out of the device stack so that they can be used with the host stack.
> 
> 
> UVC Device still works:
> 
> ```
> west build --build-dir build-uvc -b nrf52840dk/nrf52840 -S video-sw-generator samples/subsys/usb/uvc/
> ```
> 
> ![mpv-shot0001](https://private-user-images.githubusercontent.com/55351021/478803273-09162a44-6acf-4568-b1c7-e3fe0f5946cb.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI1MTM2MjgsIm5iZiI6MTc2MjUxMzMyOCwicGF0aCI6Ii81NTM1MTAyMS80Nzg4MDMyNzMtMDkxNjJhNDQtNmFjZi00NTY4LWIxYzctZTNmZTBmNTk0NmNiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA3VDExMDIwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyODIzMDM1NWE0NDBiMzY3ODRmNzY1ZTZmMDg5YWE4NjljN2Y1MmI2N2ZlYzhhNGM5YTE3MjMzMzVkMDYxYzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pUxCJ1E7qWTSbzaYnpqAaK2mB5LiTpRyACwMQ5687tQ)
> 
> All tests are done on top of [main...josuah:zephyr:pr_usb_host_class_api3](https://github.com/zephyrproject-rtos/zephyr/compare/main...josuah:zephyr:pr_usb_host_class_api3)

Hi @josuah 
I have viewed the changes in this PR and I will think about the unit test for host ecm in the future.
Furthermore, I see you also provided usbh_uvc.c in this PR. Considering the usb host video class driver is also implemented in https://github.com/zephyrproject-rtos/zephyr/pull/94085, just want to know what the plan is about the next step aimed for this class driver. Be appreciated for your any comment, Thanks.

AidenHu commented:
> > also implemented in #94085, just want to know what the plan is about the next step aimed for this class driver
> 
> The UVC implementation in my PR ("API3") is just a demo of the API, #94085 is more complete and good structure for the UVC/Video part. And needed most changes for the USB API part (as far as I understand).
> 
> I wanted to try to import all the commits of #94085 on to of #94591 so that #94085 only needs to add `usbh_uvc.c`, `CMakeLists.txt`, `Kconfig`.
> 
> I was working on nRG54LM20) (and ESP32) to be able to test UVC on DWC2 USB host. As soon as this is complete, I can try to test

@josuah, thank you for clarification.

MarkWangChinese commented on subsys/usb/host/class/usbh_uvc.c:
```diff
@@ -0,0 +1,392 @@
+
+	if (!has_vs_if) {
+		LOG_ERR("No VideoStreaming interface found");
+		return -EINVAL;
```
Should it be `ENOTSUP`?

Commit:
usb: uvc: use max host/device for MAX_FRMIVAL

Compute the maximum of the CONFIG_USBD_VIDEO_MAX_FRMIVAL and
CONFIG_USBH_VIDEO_MAX_FRMIVAL for struct definition, so that
the same struct can be relevant for both host and device
implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>
Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>
Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
usb: uvc: update header for UVC 1.5 compliance

Add input terminal descriptor and CS descriptor header and
extend frame descriptor fields (buffer size, default interval)

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>
Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
The PR https://github.com/zephyrproject-rtos/zephyr/pull/103369 got split out of this one, as it could be submitted directly on top of main.
It is now a dependency.

josuah commented on subsys/usb/host/class/usbh_uvc.c:
```diff
@@ -0,0 +1,140 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_uvc_host
```
Thank you. I imported these changes and dispatched them into the respective commits, then ran CI again.

I do not see anything in UVC device/host that requires devicetree except maybe `chosen { zephyr,camera = &uvc_host_0; };` not possible anymore.

AidenHu commented on subsys/usb/common/uvc.h:
```diff
@@ -467,7 +484,7 @@ struct uvc_frame_based_discrete_descriptor {
 	uint32_t dwMaxBitRate;
 	uint32_t dwDefaultFrameInterval;
 	uint8_t bFrameIntervalType;
```
Miss the "uint32_t dwBytesPerLine;" field below

josuah replied:
This is not the bug I thought. I will detail more here: https://github.com/zephyrproject-rtos/zephyr/pull/94085#discussion_r2752799975
There is something to fix in UVC Device.

AidenHu commented on tests/subsys/usb/uvc/prj.conf:
```diff
@@ -0,0 +1,18 @@
+
+CONFIG_UHC_DRIVER=y
+CONFIG_USB_HOST_STACK=y
+CONFIG_USBH_VIDEO_CLASS=y
```
@josuah Maybe need to add `CONFIG_POLL=y` here.

josuah commented:
Rebase on top of `main`

josuah commented on subsys/usb/common/uvc.h:
Yes, follow-up on https://github.com/zephyrproject-rtos/zephyr/pull/94085#discussion_r2752799975

carlescufi reviewed (approved):
Skeleton looks fine to me.
